src/v/cluster/archival/segment_reupload.cc

```diff
@@ -186,7 +186,7 @@ segment_collector::segment_seq segment_collector::segments() {
 void segment_collector::do_collect(segment_collector_mode mode) {
     auto projected_end_inclusive = _target_end_inclusive.value_or(
       model::offset{});
-    if (projected_end_inclusive == model::offset{}) {
+    if (projected_end_inclusive == model::offset{} || is_reupload_mode(mode)) {
         projected_end_inclusive = find_replacement_boundary(mode);
```

Comment:
This variable is named very very bad. It really is replace boundary for reupload code and end_inclusive for new segment uploads. This code is very hard to read.

Reply:
> replace boundary for reupload code

Only for compacted reuploads. For non-compacted reupload the replacement boundary is _a priori_ known. I've split some of this up a bit on my log reader branch, for whatever that's worth.